packages/flutter/lib/src/widgets/selection_arena.dart

```diff
@@ -0,0 +1,190 @@
+import 'dart:math' as math;
+
+import 'package:flutter/gestures.dart';
+import 'package:flutter/rendering.dart';
+
+import '../../widgets.dart';
+
+class SelectionArea extends StatefulWidget {
+  const SelectionArea({required this.child, Key? key}) : super(key: key);
+
+  final Widget child;
+
+  @override
+  _SelectionAreaState createState() => _SelectionAreaState();
+
+  static SelectionRegistrant? of(BuildContext context) {
+    return context.findAncestorStateOfType<_SelectionAreaState>();
+  }
+}
+
+class _SelectionState {
+  _SelectionState(this.start, this.end);
+
+  final TextPosition start;
+  final TextPosition end;
+}
+
+class _RegistrantData {
+  _RegistrantData(this.paragraph);
+
+  final RenderParagraph paragraph;
+  Rect? bounds;
+}
+
+class _SelectionAreaState extends State<SelectionArea> with ChangeNotifier implements SelectionRegistrant {
+  final GlobalKey<RawGestureDetectorState> _gestureDetectorKey = GlobalKey<RawGestureDetectorState>();
+  final Map<Type, GestureRecognizerFactory> _gestureRecognizers = <Type, GestureRecognizerFactory>{};
+  final Map<RenderParagraph, _RegistrantData> _selectionCandidates = <RenderParagraph, _RegistrantData>{};
+  final Map<RenderParagraph, _SelectionState> _selectionStates = <RenderParagraph, _SelectionState>{};
+
+  Offset? _selectionStart;
+  Offset? _selectionEnd;
+
+  @override
+  void add(RenderParagraph renderBox) {
+    _selectionCandidates[renderBox] = _RegistrantData(renderBox);
+  }
+
+  @override
+  void remove(RenderParagraph renderBox) {
+    _selectionCandidates.remove(renderBox);
+  }
+
+  @override
+  void update(RenderParagraph renderBox, Rect rect) {
+    var tl = renderBox.localToGlobal(rect.topLeft);
```

Comment:
Maybe there can be an API to let SelectionRegistrant to decide selection rect given the pointer start and end position?